src/libraries/System.DirectoryServices.Protocols/tests/DirectoryServicesProtocolsTests.cs

```diff
@@ -706,6 +707,59 @@ public void TestMultipleServerBind()
             connection.Timeout = new TimeSpan(0, 3, 0);
         }
 
+        [ConditionalFact(nameof(LdapConfigurationExists))]
+        public async Task TestAsyncSearchWithThreadRestart()
+        {
+            using LdapConnection connection = GetConnection();
+
+            string ouName = "ProtocolsGroup9";
+            string dn = "ou=" + ouName;
+
+            try
+            {
+                DeleteEntry(connection, dn);
+
+                AddOrganizationalUnit(connection, dn);
+                SearchResultEntry sre = SearchOrganizationalUnit(connection, LdapConfiguration.Configuration.SearchDn, ouName);
+                Assert.NotNull(sre);
+
+                string filter = "(objectClass=organizationalUnit)";
+                SearchRequest searchRequest = new SearchRequest(
+                                                        dn + "," + LdapConfiguration.Configuration.SearchDn,
+                                                        filter,
+                                                        SearchScope.OneLevel,
+                                                        null);
+
+                ASyncOperationState state = new ASyncOperationState(connection);
+                IAsyncResult asyncResult = connection.BeginSendRequest(
+                                                searchRequest,
+                                                PartialResultProcessing.ReturnPartialResultsAndNotifyCallback,
+                                                RunAsyncSearch,
+                                                state);
+
+                asyncResult.AsyncWaitHandle.WaitOne();
+                Assert.True(state.Exception == null, state.Exception == null ? "" : state.Exception.ToString());
+
+                // Wait for 35 seconds, so that the worker thread in PartialResultsRetriever
+                // gets terminated. Running a new async LDAP operation has to restart the worker thread.
+                await Task.Delay(35_000).ConfigureAwait(true);
```

Comment:
Please let me know if this long timeout is acceptable for the test. Another way to do this would be to have the timeout be configurable, but I'm not sure if modifying the public API should be necessary for this fix.